### ISSUE TITLE
Feature Request: Add Mattermost beside Slack

### ISSUE THREAD

**Issue by robinfrick** [metabase/metabase#2231]
It'd be nice if besides slack we could also have pulses in Mattermost.

As requested in #2158

**Comment by salsakran** [comment]
@robinfrick Do you have any experience with the Mattermost API? Is it a mostly 1-1 mapping of the Slack bot/slash command/app api structure? Would you be willing to do a bit of research to help us understand how much overlap there would be in the current way we post images to Slack with how we could post to Mattermost?

**Comment by tlrobinson** [comment]
FYI the Slack endpoints we currently use are:

* `chat.postMessage` - actually posts pulses with attached images
* `users.list` - list users for destination selection widget
* `channels.list` - list channels for destination selection widget
* `channels.create` - workaround for posting files, creates a hidden channel where we upload images
* `rtm.start` - start a streaming connection for MetaBot

**Comment by tlrobinson** [comment]
And yeah I think it would be super neat if we created a simple abstraction so it's easy for various messaging services to plug into Pulses + Metabot, but we don't want to be stuck with the lowest common denominator either. Right now all we need is:

1. list users and channels where you might want to post your pulse to
2. post messages with attached images to any of those users or channels
3. for slackbot, listen to a stream of messages on any/every channel and respond with text or pulses

But we might also want to do more in the future.

**Comment by robinfrick** [comment]
Ok as far as I can see it there is not an option for 1. The person setting up the pulse would have to type the channel name as Metabase wouldn't be able to get a list of channel/user names from mattermost.

2. & 3. look like they are supported.

For metabot:
The way integrations work on Mattermost is that you allow an integration to listen for keywords in a specific channel not everywhere at once. So that's another limitation.
http://docs.mattermost.com/developer/webhooks-outgoing.html

For pulses:
This should be better. The only limitation I see is that users have to type or copy-paste the channel names and there is no way of knowing whether it is correct without testing the pulse, so there is no immediate feedback.
http://docs.mattermost.com/developer/webhooks-incoming.html

When I wrote the request I wasn't aware of Metabot yet. I think Pulses should be relatively straight forward to adapt to Mattermosts API. I'm far from being a programmer, but I got a little bot that can post messages to mattermost up and running within very little time. I have never tried the same for slack so I can't judge, but according to mattermost's documentation at least the message format is the same. But as far as I can see it no one else +1'd my request yet so a) I can understand if you say it's not worth the effort if barely anyone wants it and b) having an incomplete experience compared to slack with metabot would probably rather raise more requests from users and possibly lead to disappointment rather than happy users.

**Comment by davidsirr** [comment]
+1 for matter most support

**Comment by pinkra** [comment]
+1 for mattermost support

**Comment by muriloht** [comment]
+1 for mattermost support

**Comment by camsaul** [comment]
Hey everyone, one way we prioritize issues is by sorting them by the number of 👍 reactions on each issue's description. So instead of leaving `+1` or 👍  comments, please just upvote by adding a reaction to the issue description.

**Comment by fernandosjp** [comment]
Just in case this issue goes foward! It could be worth looking at RocketChat as well (https://github.com/RocketChat/Rocket.Chat). Specially in case they have similar API endpoints.

On some surveys it is appearing side by side in popularity with mattermost (https://opensource.com/alternatives/slack). Today, we use Slack at the office but I'm seeing more and more of my friends going to Rocket Chat for open source alternative.

**Comment by logros** [comment]
Hi,
i use zabbix - mattermost integration via script https://github.com/muokata/zabbix-mattermost-alertscript/blob/master/mattermost.sh,

All it's ok, webhook work almost full, log from console:
Входящий вебхук получен. Содержание={"icon_url": "http://a.slack-edge.com/41b0a/img/plugins/_skeleton/service_36.png", "attachments": [ {"color": "#00ff33", "text": "OK: Много выпадающих пакетов host_pinger_ул.Александра Довженка(ул.Урицкого),5,dlink dir-100"} ], "channel": "#logros", "username": "ZABBIX Informer", "icon_emoji": ":ok_hand:"}

Messages come to chat, but icon and emoji not shown.
what's wrong with my config, please help.

**Comment by mauricio-sonegatti** [comment]
+1 for Rocket.chat 
But I think Mattermost is also a good start.

Both are the most used alternatives to Slack that are open source.
As far as I know, the pulses integration would be almost identical to Slack.

**Comment by hawaiik** [comment]
+1 for Mattermost as we just started with metabase ;)

**Comment by gsimonneaux-oceanet** [comment]
+1 for Mattermost.

**Comment by sidneiweber** [comment]
+1 for RocketChat.

**Comment by ddamasceno** [comment]
+1 for Mattermost

**Comment by mazameli** [comment]
Folks, we'd appreciate it if you could please utilize the thumbs-up reaction in the original issue description to show your support for this or any other issue, rather than `+1 for x` comments. That gives us a sortable way to see what matters most to the community, and keeps issue spam to a minimum. Cheers.

**Comment by adjenks** [comment]
@sidneiweber @gsimonneaux-oceanet @ddamasceno @hawaiik @davidsirr @muriloht,

Please delete your posts, and instead use a reaction on the initial post instead.

**Comment by narendrapsgim** [comment]
+1 for mattermost

**Comment by vapolonio** [comment]
👍

**Comment by mahenzon** [comment]
+1 for Mattermost integration

Their bot integration allows to list users, create posts and also to stream messages if needed

**Comment by adjenks** [comment]
@mouson @mahenzon @ab-smith @vapolonio @narendrapsgim 
> Please click the 👍 reaction instead of leaving a +1 comment

**Comment by petitalb** [comment]
Any update on Mattermost integration?

**Comment by nosovk** [comment]
I want to admit that Mattermost mimics to slack API. This means that current slack integration should just work, if you will allow changing a URL in it.

**Comment by paoliniluis** [comment]
@nosovk are you completely sure about this?

**Comment by nosovk** [comment]
yep, checkout the link: https://jeffschering.github.io/mmdocs/monolith/developer/api.html#slack-compatible-webhooks

**Comment by paoliniluis** [comment]
The department of defense uses Mattermost: https://docs-ironbank.dso.mil/support/

**Comment by nosovk** [comment]
It would be nice to add ai bot support also to mattermost integration (if possible)
